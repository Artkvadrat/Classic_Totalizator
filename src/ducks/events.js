import HTTPService from '../services/HTTPService';

const REQUESTED = 'events/requested';
const RECEIVED = 'events/received';
const FINISH = 'event/finish';

const requested = () => ({
  type: REQUESTED
});

const received = (data) => ({
  type: RECEIVED,
  payload: data
});

const finish = (data) => ({
  type: FINISH,
  payload: data
});

export const finishEvent = (data) => (dispatch) => {
  HTTPService.request({
    method: 'PATCH',
    path: '/api/Events/finishEvent',
    body: data
  });

  dispatch(finish(data));
};

export const getEvents = () => (dispatch) => {
  dispatch(requested());

  HTTPService.request({ path: '/api/Events/feed' }).then((data) => {
    dispatch(received(data));
  });
};

const initialState = {
  isLoading: false,
  eventsData: []
};

const eventsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED:
      return {
        ...state,
        isLoading: false,
        eventsData: [...action.payload.events]
      };

    case FINISH:
      return {
        ...state,
        eventsData: state.eventsData.map((event) => {
          if (event.id === action.payload.id) {
            return { ...event, isEnded: true };
          }

          return event;
        })
      };

    default:
      return state;
  }
};

export default eventsReducer;
