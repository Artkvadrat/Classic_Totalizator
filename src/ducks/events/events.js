import HTTPService from '../../services/HTTPService/HTTPService';

export const REQUESTED = 'events/requested';
export const RECEIVED = 'events/received';
export const FINISH = 'event/finish';

const requested = () => ({
  type: REQUESTED
});

const received = (data) => ({
  type: RECEIVED,
  payload: data
});

const finish = (data) => ({
  type: FINISH,
  payload: data
});

export const finishEvent = (data) => (dispatch) =>
  HTTPService.request({
    method: 'POST',
    path: '/api/Events/finish',
    body: data
  }).then(dispatch(finish(data)));

export const getEvents = () => (dispatch) => {
  dispatch(requested());

  return HTTPService.request({ path: '/api/Events' }).then((data) => {
    dispatch(received(data));
  });
};

const initialState = {
  isLoading: false,
  eventsData: []
};

const eventsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED:
      return {
        ...state,
        isLoading: false,
        eventsData: [...action.payload.events]
      };

    case FINISH:
      return {
        ...state,
        eventsData: state.eventsData.map((event) => {
          if (event.id === action.payload.id) {
            return { ...event, isEnded: true };
          }

          return event;
        })
      };

    default:
      return state;
  }
};

export default eventsReducer;
