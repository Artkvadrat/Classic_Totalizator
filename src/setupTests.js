import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
