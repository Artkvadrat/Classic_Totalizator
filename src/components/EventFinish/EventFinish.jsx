import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { Select, Button, Space } from 'antd';

import { finishEvent } from '../../ducks/events/events';

const EventFinish = ({ possibleResults, id }) => {
  const dispatch = useDispatch();
  const [selectedOption, setSelectedOption] = useState('Not resolved');

  const selectHandler = (value) => {
    setSelectedOption(value);
  };

  const finishEventHandler = () => {
    dispatch(finishEvent({ id, result: selectedOption }));
  };

  return (
    <Space>
      <Select
        style={{ width: 130 }}
        onChange={selectHandler}
        defaultValue={selectedOption}
        value={selectedOption}
      >
        <Select.Option value="Not resolved">Not resolved</Select.Option>
        {possibleResults.map((el) => (
          <Select.Option key={el} value={el}>
            {el}
          </Select.Option>
        ))}
      </Select>

      <Button
        disabled={selectedOption === 'Not resolved'}
        type="primary"
        onClick={finishEventHandler}
      >
        Finish
      </Button>
    </Space>
  );
};

EventFinish.propTypes = {
  id: PropTypes.string,
  possibleResults: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default EventFinish;
