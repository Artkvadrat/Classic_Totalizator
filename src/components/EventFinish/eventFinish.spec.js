import React from 'react';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';
import { finishEvent } from '../../ducks/events/events';

import '../../setupTests';

import EventFinish from './EventFinish';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn()
}));

jest.mock('../../ducks/events/events', () => ({
  finishEvent: jest.fn()
}));

describe('EventFinish component', () => {
  let wrapper;

  const dispatch = jest.fn();
  const mockedFinishEvent = jest.fn();

  const possibleResults = ['W1', 'X', 'W2'];
  const id = '1234';

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    finishEvent.mockReturnValue(mockedFinishEvent);

    wrapper = shallow(
      <EventFinish possibleResults={possibleResults} id={id} />
    );
  });

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct structure', () => {
    expect(wrapper.find('Space').length).toEqual(1);
    expect(
      wrapper.find('ForwardRef(InternalSelect)').props().defaultValue
    ).toEqual('Not resolved');
    expect(wrapper.find('Option').length).toEqual(possibleResults.length + 1);
    expect(wrapper.find('Button').length).toEqual(1);
  });

  it('should render correct values', () => {
    const optionValues = ['Not resolved', ...possibleResults];
    wrapper.find('Option').forEach((node, idx) => {
      expect(node.prop('value')).toEqual(optionValues[idx]);
    });
    expect(wrapper.find('Button').props().children).toEqual('Finish');
  });

  it('should contain button disabled on start', () => {
    expect(wrapper.find('Button').props().disabled).toEqual(true);
  });

  it('should trigger dispatch action on click button', () => {
    const button = wrapper.find('Button');
    button.simulate('click');

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(mockedFinishEvent);
  });

  it('button should become disables and vice-versa depending of options', () => {
    const select = wrapper.find('ForwardRef(InternalSelect)');

    select.simulate('change', wrapper.find('Option').first().props().value);
    expect(wrapper.find('Button').props().disabled).toEqual(true);

    wrapper
      .find('Option')
      .slice(1)
      .forEach((node) => {
        select.simulate('change', node.prop('value'));
        expect(wrapper.find('Button').props().disabled).toEqual(false);
      });
  });
});
