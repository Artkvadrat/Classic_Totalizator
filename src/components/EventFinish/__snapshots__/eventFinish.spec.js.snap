// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventFinish component should render without error 1`] = `
<Space>
  <ForwardRef(InternalSelect)
    defaultValue="Not resolved"
    onChange={[Function]}
    style={
      Object {
        "width": 130,
      }
    }
    value="Not resolved"
  >
    <Option
      value="Not resolved"
    >
      Not resolved
    </Option>
    <Option
      key="W1"
      value="W1"
    >
      W1
    </Option>
    <Option
      key="X"
      value="X"
    >
      X
    </Option>
    <Option
      key="W2"
      value="W2"
    >
      W2
    </Option>
  </ForwardRef(InternalSelect)>
  <Button
    disabled={true}
    onClick={[Function]}
    type="primary"
  >
    Finish
  </Button>
</Space>
`;
